src/rule.ts

```diff
@@ -55,12 +57,25 @@ export const rule = createRule({
 
 			scope.references.map((ref) => {
 				if (ref.resolved) {
-					const functions = ref.resolved.defs.filter(
-						(def) => def.type === DefinitionType.FunctionName,
-					) as FunctionNameDefinition[];
+					const namedFunctions: FunctionNode[] = collect(ref.resolved.defs, def => {
+						if (def.type === DefinitionType.FunctionName) {
+							return def.node;
+						}
+					});
+
+					const arrowFunctions: FunctionNode[] = collect(
+						ref.resolved.defs,
+						def => {
+							if (def.node.type === AST_NODE_TYPES.VariableDeclarator &&
+								def.node.init &&
+								def.node.init.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+								return def.node.init;
+							}
+						}
+					);
 
-					functions.forEach((fun) => {
-						const nonVoid = hasNonVoidReturnType(fun.node);
+					[...namedFunctions, ...arrowFunctions].forEach((functionNode) => {
+						const nonVoid = hasNonVoidReturnType(functionNode);
```

Comment:
I wonder if we could simplify this slightly by filtering the non-void return types with `filter` before the `forEach`? ("Simplify" is of course subjective here! But I guess I slightly prefer this as it means there's a bit less going on in the `forEach` function body).

```js
const nonVoidReturnFunctions = [...namedFunctions, ...arrowFunctions].filter(
  f => hasNonVoidReturnType(f)
);
nonVoidReturnFunctions.forEach(functionNode => {
  // ...
})
```

Reply:
Done!
It's a slightly odd bit of code because `ref.resolved.defs` gives us an array of definitions for some reason. So `nonVoidReturnFunctions` is an array but we'd only expect a single function call